### PR TITLE
[ESIMD] Use LLVM IR instead of GenX intrinsics for block loads/stores

### PR DESCRIPTION
**Pull request by v-klochkov**
This patch replaces some uses of GenX intrinsics with regular LLVM IR.

This change allowed to remove most of restrictions that previously
accompanied slm_block_load/store() and block_load/store() accepting USM ptr.
GPU RT/driver can lower 'load' instructions to one or several flat-loads
and/or gather instructions depending on the target device, alignment and length.
Same for 'store' instructions.

### CODE REVIEW

File: llvm/lib/SYCLLowerIR/ESIMD/LowerESIMD.cpp [hunk 1]
```diff
@@ -921,6 +950,46 @@ static std::string getESIMDIntrinSuffix(id::FunctionEncoding *FE,
   return Suff;
 }
 
+static void translateBlockLoad(CallInst &CI, bool IsSLM) {
+  IRBuilder<> Builder(&CI);
+
+  APInt Val =
+      parseTemplateArg(CI, 2, ESIMDIntrinDesc::GenXArgConversion::TO_I64);
```
Comment (sarnex) [inline llvm/lib/SYCLLowerIR/ESIMD/LowerESIMD.cpp:957]: could we make the `2` into a variable to explain what it is for clarity?

like `int AlignmentTemplateArgIdx = 2;`

File: llvm/lib/SYCLLowerIR/ESIMD/LowerESIMD.cpp [hunk 2]
```diff
@@ -921,6 +950,46 @@ static std::string getESIMDIntrinSuffix(id::FunctionEncoding *FE,
...
+  auto LI = Builder.CreateAlignedLoad(DataType, Op0, Align, CI.getName());
+  LI->setDebugLoc(CI.getDebugLoc());
+  CI.replaceAllUsesWith(LI);
+}
+
+static void translateBlockStore(CallInst &CI, bool IsSLM) {
+  IRBuilder<> Builder(&CI);
+
+  APInt Val =
+      parseTemplateArg(CI, 2, ESIMDIntrinDesc::GenXArgConversion::TO_I64);
```
Comment (sarnex) [inline llvm/lib/SYCLLowerIR/ESIMD/LowerESIMD.cpp:977]: same idea

File: llvm/lib/SYCLLowerIR/ESIMD/LowerESIMD.cpp [hunk 3]
```diff
@@ -921,6 +950,46 @@ static std::string getESIMDIntrinSuffix(id::FunctionEncoding *FE,
...
+static void translateBlockLoad(CallInst &CI, bool IsSLM) {
+  IRBuilder<> Builder(&CI);
+
+  APInt Val =
+      parseTemplateArg(CI, 2, ESIMDIntrinDesc::GenXArgConversion::TO_I64);
+  MaybeAlign Align(Val.getZExtValue());
+
+  auto Op0 = CI.getArgOperand(0);
+  auto DataType = CI.getType();
+  if (IsSLM) {
```
Comment (sarnex) [inline llvm/lib/SYCLLowerIR/ESIMD/LowerESIMD.cpp:962]: can you test with opaque pointers and make sure this inttoptr conversion works? maybe the types are different

File: sycl/include/sycl/ext/intel/esimd/memory.hpp [hunk 4]
```diff
@@ -279,36 +266,23 @@ scatter(Tx *p, Toffset offset, simd<Tx, N> vals, simd_mask<N> mask = 1) {
...
 /// @param addr The address to load from.
 /// @param Flags Specifies the alignment.
 /// @return A vector of loaded elements.
 ///
-template <typename Tx, int N, typename Flags = vector_aligned_tag,
-          class T = detail::__raw_t<Tx>,
-          typename = std::enable_if_t<is_simd_flag_type_v<Flags>>>
-__ESIMD_API simd<Tx, N> block_load(const Tx *addr, Flags = {}) {
-  constexpr unsigned Sz = sizeof(T) * N;
-  static_assert(Sz >= detail::OperandSize::OWORD,
```
Comment (sarnex) [inline sycl/include/sycl/ext/intel/esimd/memory.hpp:296]: it's really cool we can remove all these restrictions, but are we sure the hardware can support them? i assume IGC lowers the loads to the same hardware instructions as the VC intrinsic we used before, so i would think the same restrictions apply, but i am not a hardware expert.

File: sycl/test/esimd/intrins_trans.cpp [hunk 5]
```diff
@@ -196,7 +197,7 @@ SYCL_EXTERNAL vec<unsigned short, 8> get8ui16() SYCL_ESIMD_FUNCTION;
 SYCL_EXTERNAL vec<unsigned char, 8> get8ui8() SYCL_ESIMD_FUNCTION;
 
 SYCL_EXTERNAL void
-test_mem_intrins(uint64_t addr, const vec<float, 8> &xf,
```
Comment (sarnex) [inline sycl/test/esimd/intrins_trans.cpp:199]: why the change here? are we sure we nothing regressed?

File: sycl/test-e2e/ESIMD/esimd_test_utils.hpp [hunk 6]
```diff
@@ -610,4 +610,37 @@ std::string toString(sycl::ext::intel::experimental::esimd::lsc_data_size DS) {
   return "INVALID lsc_data_size";
 }
 
+template <typename... ArgT> void printTestLabel(queue Q, ArgT &&...Args) {
+  auto Dev = Q.get_device();
+  auto Name = Dev.get_info<sycl::info::device::name>();
+  auto Driver = Dev.get_info<sycl::info::device::driver_version>();
+  std::cout << "Running on " << Name << ", driver=[" << Driver << "]";
+  if constexpr (sizeof...(ArgT) > 0) {
```
Comment (sarnex) [inline sycl/test-e2e/ESIMD/esimd_test_utils.hpp:618]: cool utilities

File: sycl/include/sycl/ext/intel/esimd/detail/memory_intrin.hpp [hunk 7]
```diff
@@ -38,6 +38,12 @@ const std::array<__ESIMD_NS::rgba_channel, 4> ChannelMaskArray{
 
 namespace sycl {
 __SYCL_INLINE_VER_NAMESPACE(_V1) {
+
+namespace ext::intel::esimd {
+template <typename AccessorTy>
+__ESIMD_API SurfaceIndex get_surface_index(AccessorTy acc);
```
[the remark this answers is not in the document]
Comment (v-klochkov) [reply sycl/include/sycl/ext/intel/esimd/detail/memory_intrin.hpp:44]: This is pre-declaration. The body is defined in memory.hpp, but is needed in memory_intrin.hpp and memory_intrin.hpp cannot include memory.hpp

File: sycl/test/esimd/intrins_trans.cpp (hunk 5 above)
Comment (v-klochkov) [reply sycl/test/esimd/intrins_trans.cpp:199]: 'addr' was used only to test __esimd_svm_block_ld/st and __esimd_svm_block_ld_unaligned. Those intrinsics accepted uint64_t.
This PR replaced/changed those intrinsics with intrinsics accepting pointers. Thus this change was needed.

File: llvm/lib/SYCLLowerIR/ESIMD/LowerESIMD.cpp [hunk 8]
```diff
@@ -823,8 +829,31 @@ static APInt parseTemplateArg(id::FunctionEncoding *FE, unsigned int N,
...
+                              ESIMDIntrinDesc::GenXArgConversion Conv =
+                                  ESIMDIntrinDesc::GenXArgConversion::NONE) {
+  Function *F = CI.getCalledFunction();
+  llvm::esimd::assert_and_diag(F, "function to translate is invalid");
+
+  StringRef MnglName = F->getName();
+  using Demangler = id::ManglingParser<SimpleAllocator>;
+  Demangler Parser(MnglName.begin(), MnglName.end());
+  id::Node *AST = Parser.parse();
+  llvm::esimd::assert_and_diag(
```
[the remark this answers is not in the document]
Comment (v-klochkov) [reply llvm/lib/SYCLLowerIR/ESIMD/LowerESIMD.cpp:848]: I tried separating that common code to a function - it turned out to be not so simple, or maybe even not doable.
If I understood it correctly, 'AST' is correct only while 'Parser' is alive. Separating that code into a function will produce 'AST' and cast of it to 'id::FunctionEncoding*', but will invalidate it after exit from function, which will make the returned AST not usable.

File: llvm/lib/SYCLLowerIR/ESIMD/LowerESIMD.cpp [hunk 9]
```diff
@@ -921,6 +950,46 @@ static std::string getESIMDIntrinSuffix(id::FunctionEncoding *FE,
...
+  IRBuilder<> Builder(&CI);
+
+  APInt Val =
+      parseTemplateArg(CI, 2, ESIMDIntrinDesc::GenXArgConversion::TO_I64);
+  MaybeAlign Align(Val.getZExtValue());
+
+  auto Op0 = CI.getArgOperand(0);
+  auto DataType = CI.getType();
+  if (IsSLM) {
+    // Convert 'uint32_t' to 'addrspace(3)*' pointer.xxyy
```
[the remark this answers is not in the document]
Comment (v-klochkov) [reply llvm/lib/SYCLLowerIR/ESIMD/LowerESIMD.cpp]: It is one of several temporary text marks/labels to navigate/jump to them. Removed in 1e722d1af21d4d979f9e352e053a0ae4d340e45a

File: llvm/lib/SYCLLowerIR/ESIMD/LowerESIMD.cpp (hunk 1 above)
Comment (v-klochkov) [reply llvm/lib/SYCLLowerIR/ESIMD/LowerESIMD.cpp]: Fixed in 1e722d1af21d4d979f9e352e053a0ae4d340e45a

File: llvm/lib/SYCLLowerIR/ESIMD/LowerESIMD.cpp (hunk 2 above)
Comment (v-klochkov) [reply llvm/lib/SYCLLowerIR/ESIMD/LowerESIMD.cpp]: Fixed in 1e722d1af21d4d979f9e352e053a0ae4d340e45a

File: llvm/lib/SYCLLowerIR/ESIMD/LowerESIMD.cpp (hunk 3 above)
Comment (v-klochkov) [reply llvm/lib/SYCLLowerIR/ESIMD/LowerESIMD.cpp:963]: Yes, I did the tests.
Also, added few more tests (1 check to existing LIT test) and added 1 E2E test in 1e722d1af21d4d979f9e352e053a0ae4d340e45a

Thank you for this comment. The additional testing showed some misses in the initial version of the patch.

File: sycl/include/sycl/ext/intel/esimd/memory.hpp (hunk 4 above)
Comment (v-klochkov) [reply sycl/include/sycl/ext/intel/esimd/memory.hpp:296]: I extended the tests.
IGC lowers VC intrinsics and LLVM-IR load/store instructions differently.
VC-intrinsics are lowered mostly 1-to-1.

File: sycl/test-e2e/ESIMD/esimd_test_utils.hpp (hunk 6 above)
Comment (v-klochkov) [reply sycl/test-e2e/ESIMD/esimd_test_utils.hpp:618]: Thank you )

File: llvm/lib/SYCLLowerIR/ESIMD/LowerESIMD.cpp [hunk 10]
```diff
@@ -805,8 +811,8 @@ static APInt parseTemplateArg(id::FunctionEncoding *FE, unsigned int N,
     const std::string_view &TyStr = ValL->getType();
     if (Conv == ESIMDIntrinDesc::GenXArgConversion::NONE && TyStr.size() != 0)
       // Overwrite Ty with IntegerLiteral's size
-      Ty =
-          parsePrimitiveTypeString(StringRef(&*TyStr.begin(), TyStr.size()), Ctx);
+      Ty = parsePrimitiveTypeString(StringRef(&*TyStr.begin(), TyStr.size()),
```
[the remark this answers is not in the document]
Comment (v-klochkov) [reply llvm/lib/SYCLLowerIR/ESIMD/LowerESIMD.cpp:814]: It is rather exception here.
clang-format produced lots of good changes to lines modified by me and only 1 or 2 changes unrelated to the main PR. In this particular case I decided to keep those 1-or-2 unrelated clang-format fixes instead of searching for it and cancelling.

Normally, if there are many unrelated clang-format changes, I cancel them an apply only required for modified lines.

File: llvm/lib/SYCLLowerIR/ESIMD/LowerESIMD.cpp (hunk 8 above)
[the remark this answers is not in the document]
Comment (sarnex) [reply llvm/lib/SYCLLowerIR/ESIMD/LowerESIMD.cpp:848]: ah ok, i didnt know that, thanks

File: sycl/include/sycl/ext/intel/esimd/memory.hpp (hunk 4 above)
Comment (sarnex) [reply sycl/include/sycl/ext/intel/esimd/memory.hpp:296]: oh ok, interesting

File: sycl/test/esimd/intrins_trans.cpp (hunk 5 above)
Comment (sarnex) [reply sycl/test/esimd/intrins_trans.cpp:199]: got it, thx

File: llvm/lib/SYCLLowerIR/ESIMD/LowerESIMD.cpp (hunk 10 above)
[the remark this answers is not in the document]
Comment (turinevgeny) [reply llvm/lib/SYCLLowerIR/ESIMD/LowerESIMD.cpp:814]: @v-klochkov, btw, do you apply clang-format to the whole file, or to the diff? If applied to the diff, these issues will not arise. Can share a command if needed.

File: sycl/include/sycl/ext/intel/esimd/memory.hpp [hunk 11]
```diff
@@ -262,14 +262,17 @@ scatter(Tx *p, Toffset offset, simd<Tx, N> vals, simd_mask<N> mask = 1) {
   scatter<Tx, N>(p, simd<Toffset, N>(offset), vals, mask);
 }
 
-/// Loads a contiguous block of memory from given memory address and returns
-/// the loaded data as a vector. Actual code generated depends on the
-/// alignment parameter.
+/// Loads a contiguous block of memory from the given memory address \p addr
+/// and returns the loaded data as a vector.
+/// The generated code depends on the combination {T, N, Flags}.
+/// Providing flags specifying the alignment of 16-bites or more produces more
```
Comment (sarnex) [inline sycl/include/sycl/ext/intel/esimd/memory.hpp:268]: bites->bytes, same typo in a few spots

**Review by sarnex — APPROVED** [review]
looks good, thanks!

**Review by turinevgeny — APPROVED** [review]
LGTM.

File: llvm/lib/SYCLLowerIR/ESIMD/LowerESIMD.cpp [hunk 12]
```diff
@@ -154,6 +154,7 @@ struct ESIMDIntrinDesc {
     TO_I8,  // convert vector of N-bit integer to 18-bit
```
Comment (vmustya) [inline llvm/lib/SYCLLowerIR/ESIMD/LowerESIMD.cpp:154]:
```suggestion
    TO_I8,  // convert vector of N-bit integer to 8-bit
```

File: sycl/include/sycl/ext/intel/esimd/memory.hpp (hunk 11 above)
Comment (v-klochkov) [reply sycl/include/sycl/ext/intel/esimd/memory.hpp]: Fixed in 2241ceea91cbed7c4dbbcafa0ef201fe706a8097. Thank you

**Comment by v-klochkov** [comment]
@turinevgeny , @fineg74 - please re-review this additional patch: 5c876ee5da8a2910d39e7b4e3ec4b5dd9c77d653
It replaces 2 old routines (minLinuxDriver() and minWindowsDriver() with a new isGPUDriverGE()  // GE stands for Greater or Equal.

**Comment by turinevgeny** [comment]
> @turinevgeny , @fineg74 - please re-review this additional patch: [5c876ee](https://github.com/intel/llvm/commit/5c876ee5da8a2910d39e7b4e3ec4b5dd9c77d653) It replaces 2 old routines (minLinuxDriver() and minWindowsDriver() with a new isGPUDriverGE() // GE stands for Greater or Equal. The motivation for doing that were fails with windows/opencl and the new fails on new Gen12-CI.

Looks okay.

**Review by turinevgeny — APPROVED** [review]
